alphacsc/_solver_d.py

```diff
@@ -0,0 +1,692 @@
+                                  adaptive_step_size=adaptive_step_size,
+                                  verbose=verbose, debug=debug,
+                                  name="Update v")
+            uv_hat = np.c_[u_hat, v_hat]
+            if debug:
+                pobj.extend(pobj_v)
+
+        if self.window:
+            uv_hat[:, n_channels:] *= tukey_window_
+
+        if debug:
+            return uv_hat, pobj
+        return uv_hat
+
+
+class DSolver(BaseDSolver):
+    """A class for 'fista' solver_d when rank1 is False. """
+
+    def __init__(self,
+                 solver_d,
+                 rank1,
+                 uv_constraint,
+                 window,
+                 eps,
+                 max_iter,
+                 momentum,
+                 random_state):
+
+        super().__init__(solver_d,
+                         rank1,
+                         uv_constraint,
+                         window,
+                         eps,
+                         max_iter,
+                         momentum,
+                         random_state)
+
+    def _window(self, D_hat0, n_times_atom):
+        tukey_window_ = None
+
+        if self.window:
+            tukey_window_ = tukey_window(n_times_atom)[None, None, :]
+            D_hat0 = D_hat0 / tukey_window_
+
+        return D_hat0, tukey_window_
+
+    def _objective(self, D, z_encoder, tukey_window_, full=False):
+
+        def objective(D, full=False):
+            if self.window:
+                D = D * tukey_window_
+
+            if z_encoder.loss == 'l2':
+                return compute_objective(D=D,
+                                         constants=z_encoder.get_constants())
+
+            return compute_X_and_objective_multi(z_encoder.X,
+                                                 z_encoder.get_z_hat(),
+                                                 D_hat=D,
+                                                 loss=z_encoder.loss,
+                                                 loss_params=z_encoder.loss_params)  # noqa
+
+        return objective
+
+    def get_max_error_dict(self, z_encoder):
+        """Get the maximal reconstruction error patch from the data as a new atom
+
+        This idea is used for instance in [Yellin2017]
+
+        Parameters
+        ----------
+        z_encoder : BaseZEncoder
+            ZEncoder object to be able to compute the largest error patch.
+
+        Return
+        ------
+        dk: array, shape (n_channels, n_times_atom,)
+            New atom for the dictionary, chosen as the chunk of data with the
+            maximal reconstruction error.
+
+        [Yellin2017] BLOOD CELL DETECTION AND COUNTING IN HOLOGRAPHIC LENS-FREE
+        IMAGING BY CONVOLUTIONAL SPARSE DICTIONARY LEARNING AND CODING.
+        """
+        n_times_atom = z_encoder.n_times_atom
+        d0 = z_encoder.get_max_error_patch()
+
+        if self.window:
+            d0 = d0 * tukey_window(n_times_atom)[None, :]
```

Comment:
It would be nice to compute this window once for all and store it in the class no?